hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/JoinMessage.java

```diff
@@ -20,16 +20,23 @@
 import com.hazelcast.nio.ObjectDataInput;
 import com.hazelcast.nio.ObjectDataOutput;
 import com.hazelcast.nio.serialization.IdentifiedDataSerializable;
+import com.hazelcast.version.Version;
 
 import java.io.IOException;
 import java.util.ArrayList;
 import java.util.Collection;
 import java.util.Collections;
 
+// Used as request and response in join protocol
 public class JoinMessage implements IdentifiedDataSerializable {
 
     protected byte packetVersion;
     protected int buildNumber;
+    /**
+     * this is populated with the codebase version of the node trying to join the cluster
+     * (ie {@link com.hazelcast.instance.Node#getVersion()}).
+     */
+    protected Version version;
     protected Address address;
```

Comment:
`JoinMessage` is used as:
* payload in `MasterDiscoveryOperation`. A not joined node configured with `TcpIpJoiner` sends this out to a list of possible member addresses (deducted from tcp-ip configuration) to discover who is master and then join.
* response from a cluster's master node to a node sending a `JoinRequest` when using the multicast joiner; this is the way a not-joined node becomes aware of the cluster master and then sends out a `JoinRequestOperation` to actually join the cluster.
* request & response payload in `JoinCheckOperation` when searching for other clusters in the context of split-brain handling.

`JoinRequest` is used to join a cluster from either multicast or TCP/IP joiner, as well as from multicast joiner when searching for other clusters.

I have already added javadoc to important methods in `ClusterJoinManager` in an attempt to keep track of what each method does. It's not complete documentation of the discovery & join process, but should help.